eclipse-theia/theia
# Factory for menu nodes

This PR introduces a MenuModelFactory to provide complete control over the menu nodes in downstream projects:
 - A downstream project can show an ampersand (`&`) in the electron menu label. By default, they're not rendered and kept for accelerators on Windows: https://github.com/electron/electron/issues/9584,
 - A Downstream project can customize the menu order. Instead of always specifying a `sortString` option, the menu node itself can provide better sorting.

Before:
![Screen Shot 2021-03-22 at 15 23 39](https://user-images.githubusercontent.com/1405703/112009025-5110b680-8b26-11eb-81f9-d1145e7d08ec.jpg)


After:
![Screen Shot 2021-03-22 at 15 43 13](https://user-images.githubusercontent.com/1405703/112009000-4b1ad580-8b26-11eb-95a9-8c8597579de1.jpg)


#### What it does


#### How to test


 - Core tests should pass,
 - Check the `Sample Menu` > `Foo Submenu`,
   - submenu ordering should ignore case sensitivity,
   - you should see the ampersand (`&`) in the menu label when running the electron example.

This PR does not change the default behavior.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
